File: RaftSwift/Sources/RaftGRPCTransport/RaftGRPCServer.swift
import GRPCCore
import GRPCNIOTransportHTTP2
import Logging
import RaftCore

/// A Raft server that uses gRPC for communication
public final class RaftGRPCServer: RaftNodeApplication {
    public let ownPeer: Peer
    public let peers: [Peer]
    public let persistence: any RaftNodePersistence
    public let collectMetrics: Bool
    public let useManualLock: Bool

    /// The logger
    let logger = Logger(label: "raft.RaftGRPCServer")

    public init(
        ownPeer: Peer,
        peers: [Peer],
        persistence: any RaftNodePersistence,
        collectMetrics: Bool,
        useManualLock: Bool
    ) {
        self.ownPeer = ownPeer
        self.peers = peers
        self.persistence = persistence
        self.collectMetrics = collectMetrics
        self.useManualLock = useManualLock
    }

    public func serve() async throws {
        let node: any RaftNodeProtocol = if useManualLock {
            RaftNodeManualLock(
                ownPeer,
                peers: peers,
                config: RaftConfig(),
                transport: GRPCNodeTransport(clientPool: GRPCClientPool(interceptors: [
                    ServerIDInjectionInterceptor(peerID: ownPeer.id),
                ])),
                persistence: persistence,
                collectMetrics: collectMetrics,
            )
        } else {
            RaftNode(
                ownPeer,
                peers: peers,
                config: RaftConfig(),
                transport: GRPCNodeTransport(clientPool: GRPCClientPool(interceptors: [
                    ServerIDInjectionInterceptor(peerID: ownPeer.id),
                ])),
                persistence: persistence,
                collectMetrics: collectMetrics,
            )
        }
        let peerService = PeerService(node: node)
        let clientService = ClientService(node: node)

        let partitionInterceptor = NetworkPartitionInterceptor(logger: logger)
        let partitionService = PartitionService(partitionController: partitionInterceptor)

        let server: GRPCServer<HTTP2ServerTransport.Posix> = GRPCServer(
            transport: .http2NIOPosix(
                address: .ipv4(host: "0.0.0.0", port: ownPeer.port),
                transportSecurity: .plaintext,
                config: HTTP2ServerTransport.Posix.Config.defaults(configure: { config in
                    config.rpc.maxRequestPayloadSize = 100 * 1024 * 1024 // 100 MB
                }),
            ),
            services: [
                peerService,
                clientService,
                partitionService,
            ],
            interceptors: [
                partitionInterceptor,
            ],
        )

        try await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask {
                try await server.serve()
            }

            if let address = try await server.listeningAddress {
                logger.info("Server listening on \(address)")
            }

            logger.info("Starting node")
            await node.start()
        }
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/transport/grpc/RaftGRPCServer.kt
package raft.transport.grpc

import io.grpc.netty.NettyServerBuilder
import io.grpc.Server
import kotlinx.coroutines.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import raft.core.node.RaftNode
import raft.core.node.RaftNodeApplication
import raft.core.node.RaftNodePersistence
import raft.core.utils.types.Peer
import raft.core.utils.types.RaftConfig
import raft.transport.grpc.services.ClientService
import raft.transport.grpc.services.PartitionService
import raft.transport.grpc.services.PeerService
import raft.transport.grpc.utils.GRPCClientPool
import java.net.InetAddress
import java.net.InetSocketAddress

/**
 * A Raft server that uses gRPC for communication.
 */
class RaftGRPCServer(
    ownPeer: Peer,
    peers: MutableList<Peer>,
    persistence: RaftNodePersistence
) : RaftNodeApplication(ownPeer, peers, persistence) {

    private val logger: Logger = LoggerFactory.getLogger(RaftGRPCServer::class.java)
    private var server: Server? = null

    override fun serve() {
        runBlocking {
            serveAsync()
        }
    }

    /**
     * Starts the server asynchronously.
     */
    suspend fun serveAsync() {
        val node = RaftNode(
            ownPeer = ownPeer,
            peers = peers,
            config = RaftConfig(),
            transport = GRPCNodeTransport(
                clientPool = GRPCClientPool(
                    interceptors = listOf(
                        ServerIDInjectionInterceptor(peerID = ownPeer.id)
                    )
                )
            ),
            persistence = persistence
        )

        val peerService = PeerService(node)
        val clientService = ClientService(node)
        val partitionInterceptor = NetworkPartitionInterceptor()
        val partitionService = PartitionService(partitionInterceptor)

        server = NettyServerBuilder.forAddress(
            InetSocketAddress(InetAddress.getByName(ownPeer.address), ownPeer.port)
        )
            .addService(peerService)
            .addService(clientService)
            .addService(partitionService)
            .intercept(partitionInterceptor)
            .build()

        server?.start()
        logger.info("Server listening on ${ownPeer.address}:${ownPeer.port}")

        // Start the Raft node (won't be blocked by awaitTermination)
        val nodeJob = CoroutineScope(Dispatchers.Default).launch {
            logger.info("Starting node")
            node.start()
        }

        // Wait for the server to terminate (in blocking context)
        withContext(Dispatchers.IO) {
            server?.awaitTermination()
        }

        // Optionally, wait for node job to complete or cancel it when server shuts down
        nodeJob.cancelAndJoin()
    }

    /**
     * Shuts down the server gracefully.
     */
    suspend fun shutdown() {
        withContext(Dispatchers.IO) {
            server?.shutdown()
            server?.awaitTermination()
        }
        logger.info("Server shut down")
    }
}